src/gretel_synthetics/generate.py

```diff
@@ -7,106 +7,32 @@
 
-PredString = namedtuple("pred_string", ["data"])
-
-
-@dataclass
-class gen_text:
-    """
-    A record that is yielded from the ``generate_text`` generator.
-
-    Attributes:
-        valid: True, False, or None. If the line passed a validation function,
-            then this will be ``True``. If the validation function raised an exception
-            then this will be automatically set to ``False``. If no validation function
-            is used, then this value will be ``None.``
-        text: The actual record as a string
-        explain: A string that describes why a record failed validation. This is the
-            string representation of the ``Exception`` that is thrown in a validation
-            function. This will only be set if validation fails, otherwise will be ``None.``
-        delimiter: If the generated text are column/field based records. This will hold the delimiter
-            used to separate the fields from each other.
-    """
-
-    valid: bool = None
-    text: str = None
-    explain: str = None
-    delimiter: str = None
-
-    def as_dict(self) -> dict:
-        """Serialize the generated record to a dictionary
-        """
-        return asdict(self)
-
-    def values_as_list(self) -> List[str]:
-        """Attempt to split the generated text on the provided delimiter
-
-        Returns:
-            A list of values that are separated by the object's delimiter or None is there
-            is no delimiter in the text
-        """
-        if self.delimiter is not None:
-            tmp = self.text.rstrip(self.delimiter)
-            return tmp.split(self.delimiter)
-        return None
 
+gen_text = generator.gen_text
+PredString = generator.PredString
 
 logging.basicConfig(
     format="%(asctime)s : %(threadName)s : %(levelname)s : %(message)s",
     level=logging.INFO,
 )
 
 
-def _load_tokenizer(store: "BaseConfig") -> spm.SentencePieceProcessor:
-    logging.info("Loading SentencePiece tokenizer")
-    sp = spm.SentencePieceProcessor()
-    sp.Load(store.tokenizer_model)
-    return sp
-
-
-def _prepare_model(
-    sp: spm, batch_size: int, store: "BaseConfig"
-) -> tf.keras.Sequential:  # pragma: no cover
-    model = _build_sequential_model(
-        vocab_size=len(sp), batch_size=batch_size, store=store
-    )
-
-    load_dir = store.checkpoint_dir
-
-    model.load_weights(tf.train.latest_checkpoint(load_dir)).expect_partial()
-
-    model.build(tf.TensorShape([1, None]))
-    model.summary()
-
-    return model
-
-
-def _load_model(
-    store: "BaseConfig",
-) -> Tuple[spm.SentencePieceProcessor, tf.keras.Sequential]:
-    sp = _load_tokenizer(store)
-    model = _prepare_model(sp, 1, store)
-    return sp, model
-
-
 def generate_text(
-    config: "BaseConfig",
+    config: "LocalConfig",
     start_string: str = "<n>",
     line_validator: Callable = None,
     max_invalid: int = 1000,
-    num_lines: int = None
+    num_lines: int = None,
+    parallelism: int = 1,
```

Comment:
Wondering if we should default to `0` here which defaults us to max CPUs.  @zredlined thoughts?

Reply:
I realized that this means I need to set it to `1` explicitly in `test_generate.py` in every invocation, because the subprocess-based parallelism doesn't work too nicely with mocks

Reply:
Left a review below- by default it looks like we'll still use a GPU if available for text generation, where I think we'll want to default to max CPUs, right?

Reply:
Ah, yeah, I pushed a change that will set `CUDA_VISIBLE_DEVICES` to `-1` for every spawned worker process. This should ensure reliably that the GPU isn't used in the worker processes (need to verify, if not, looks like the option is to enumerate devices and enclose the code in a `with tf.device(...)` block). This would mean:
- `parallelism=1`: no worker processes are used, so default settings for GPU/CPU remain
- otherwise, spawn parallel worker processes which use CPU only.
If I understand you correctly you are fine with `parallelism != 1` implying CPU-only, but want an option added that will force CPU for text generation (regardless of any global settings that could be imposed via the environment and thus also affect training) even in non-parallel execution mode?

Reply:
I think the env change in spawned processes only effectively addresses all these issues well.